lib/widgets/emoji_reaction.dart

```diff
@@ -360,3 +366,225 @@ class _TextEmoji extends StatelessWidget {
       text);
   }
 }
+
+
+void showEmojiPickerSheet({required BuildContext pageContext, required Message message}) {
+  final store = PerAccountStoreWidget.of(pageContext);
+
+  showModalBottomSheet<void>(
+    context: pageContext,
+    // Clip.hardEdge looks bad; Clip.antiAliasWithSaveLayer looks pixel-perfect
+    // on my iPhone 13 Pro but is marked as "much slower":
+    //   https://api.flutter.dev/flutter/dart-ui/Clip.html
+    clipBehavior: Clip.antiAlias,
+    useSafeArea: true,
+    isScrollControlled: true,
+    builder: (BuildContext context) {
+      return SafeArea(
+        // For _EmojiPickerItem, and RealmContentNetworkImage used in ImageEmojiWidget.
+        child: PerAccountStoreWidget(
+          accountId: store.accountId,
+          child: Padding(
+            // By default, when software keyboard is opened, the ListView
+            // expands behind the software keyboard — resulting in some
+            // list entries being covered by the keyboard. Add explicit
+            // bottom padding the size of the keyboard, which fixes this.
+            padding: EdgeInsets.only(bottom: MediaQuery.viewInsetsOf(context).bottom),
+            child: EmojiPicker(pageContext: pageContext, message: message))));
+    });
+}
+
+class EmojiPicker extends StatefulWidget {
+  const EmojiPicker({
+    super.key,
+    required this.pageContext,
+    required this.message,
+  });
+
+  final BuildContext pageContext;
+  final Message message;
+
+  @override
+  State<EmojiPicker> createState() => _EmojiPickerState();
+}
+
+class _EmojiPickerState extends State<EmojiPicker> with PerAccountStoreAwareStateMixin<EmojiPicker> {
+  late TextEditingController _controller;
+
+  EmojiAutocompleteView? _viewModel;
+  List<EmojiAutocompleteResult> _resultsToDisplay = const [];
+
+  @override
+  void initState() {
+    super.initState();
+    _controller = TextEditingController();
+  }
+
+  @override
+  void onNewStore() {
+    final store = PerAccountStoreWidget.of(context);
+
+    if (_viewModel != null) {
+      _controller.removeListener(_handleViewModelUpdate);
+      _viewModel!.dispose();
+      _resultsToDisplay = const [];
+    }
+    _viewModel = EmojiAutocompleteView.init(store: store, query: EmojiAutocompleteQuery(''))
+      ..addListener(_handleViewModelUpdate);
+    _controller.addListener(_handleControllerUpdate);
```

Comment:
This can go in `initState` (like in `AutocompleteField`), so the listener gets added only once even if the store gets replaced.

The listener won't be called before this point anyway — it won't be called until there's that TextField attached to the controller, which happens only after this is mounted and built, and `onNewStore` will get called after the state is mounted and before it's built.